### PR TITLE
fix mypy typing errors in pytorch_lightning.utilities.data.py

### PR DESCRIPTION
**Pull request by nandwalritik**
## What does this PR do?
* mypy errors
```src/pytorch_lightning/utilities/data.py:46: error: Cannot resolve name "BType" (possible cyclic definition)  [misc]
src/pytorch_lightning/utilities/data.py:67: error: Incompatible types in "yield" (actual type "None", expected type "int")  [misc]
src/pytorch_lightning/utilities/data.py:109: error: Argument 1 to "len" has incompatible type "Union[DataLoader[Any], Iterable[Any]]"; expected "Sized"  [arg-type]
src/pytorch_lightning/utilities/data.py:119: error: Argument 1 to "has_iterable_dataset" has incompatible type "Union[DataLoader[Any], Iterable[Any]]"; expected "DataLoader[Any]"  [arg-type]
src/pytorch_lightning/utilities/data.py:131: error: Name "pl.Strategy" is not defined  [name-defined]
src/pytorch_lightning/utilities/data.py:138: error: Item "LightningDataModule" of "Union[LightningModule, LightningDataModule]" has no attribute "device"  [union-attr]
src/pytorch_lightning/utilities/data.py:190: error: Argument 2 to "_get_dataloader_init_args_and_kwargs" has incompatible type "Union[Sampler[Any], Iterable[Any]]"; expected "Optional[Sampler[Any]]"  [arg-type]
src/pytorch_lightning/utilities/data.py:241: error: Cannot access "__init__" directly  [misc]
src/pytorch_lightning/utilities/data.py:283: error: Incompatible types in assignment (expression has type "List[str]", variable has type "Set[str]")  [assignment]
src/pytorch_lightning/utilities/data.py:298: error: Incompatible types in assignment (expression has type "List[Any]", variable has type "Set[Any]")  [assignment]
src/pytorch_lightning/utilities/data.py:417: error: Argument 1 to "FastForwardSampler" has incompatible type "Optional[Sampler[Any]]"; expected "Union[Sampler[Any], Generator[Any, Any, Any]]"  [arg-type]
src/pytorch_lightning/utilities/data.py:534: error: Argument 1 to "get_len" has incompatible type "Dataset[Any]"; expected "DataLoader[Any]"  [arg-type]
Found 12 errors in 1 file (checked 240 source files)
```

Fix mypy errors attributed to pytorch_lightning.utilities.data.py for issue  #13445 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Yeah

### CODE REVIEW

File: src/pytorch_lightning/utilities/data.py [hunk 1]
```diff
@@ -264,9 +266,9 @@ def _get_dataloader_init_args_and_kwargs(
         dl_kwargs["batch_sampler"] = None
         dl_kwargs["sampler"] = None
     else:
-        dl_kwargs.update(_dataloader_init_kwargs_resolve_sampler(dataloader, sampler, mode=mode))
+        dl_kwargs.update(_dataloader_init_kwargs_resolve_sampler(dataloader, sampler, mode=mode))  # type: ignore
```
Comment (otaj) [inline src/pytorch_lightning/utilities/data.py:269]: I think this could be easily typed without using `# type: ignore`

File: src/pytorch_lightning/utilities/data.py [hunk 2]
```diff
@@ -309,7 +311,7 @@ def _get_dataloader_init_args_and_kwargs(
 
 
 def _dataloader_init_kwargs_resolve_sampler(
-    dataloader: DataLoader, sampler: Optional[Sampler], mode: Optional[RunningStage] = None
+    dataloader: DataLoader, sampler: Union[Sampler, Generator], mode: Optional[RunningStage] = None
```
Comment (otaj) [inline src/pytorch_lightning/utilities/data.py:314]: How do we get a `Generator` in here?

File: src/pytorch_lightning/utilities/data.py [hunk 3]
```diff
@@ -127,14 +127,16 @@ def has_len(dataloader: Union[DataLoader, Iterable]) -> bool:
 
 def has_len_all_ranks(
     dataloader: DataLoader,
-    training_type: "pl.Strategy",
+    training_type: "pl.Strategy",  # type: ignore
```
Comment (otaj) [inline src/pytorch_lightning/utilities/data.py:130]: why do we ignore here?

File: src/pytorch_lightning/utilities/data.py [hunk 4]
```diff
@@ -127,14 +127,16 @@ def has_len(dataloader: Union[DataLoader, Iterable]) -> bool:
...
     model: Union["pl.LightningModule", "pl.LightningDataModule"],
 ) -> bool:
     """Checks if a given Dataloader has ``__len__`` method implemented i.e. if it is a finite dataloader or
     infinite dataloader."""
     try:
         local_length = len(dataloader)
-        total_length = training_type.reduce(torch.tensor(local_length).to(model.device), reduce_op="sum")
+        total_length = training_type.reduce(
+            torch.tensor(local_length).to(model.device), reduce_op="sum"
+        )  # type: ignore
```
Comment (otaj) [inline src/pytorch_lightning/utilities/data.py:139]: Why to ignore here?

File: src/pytorch_lightning/utilities/data.py [hunk 5]
```diff
@@ -234,7 +236,7 @@ def _get_dataloader_init_args_and_kwargs(
         arg_names = ()
 
     # get the dataloader instance `__init__` parameters
-    params = dict(inspect.signature(dataloader.__init__).parameters)
+    params = dict(inspect.signature(dataloader.__init__).parameters)  # type: ignore
```
Comment (otaj) [inline src/pytorch_lightning/utilities/data.py:239]: This could also be easily typed

File: src/pytorch_lightning/utilities/data.py [hunk 6]
```diff
@@ -408,15 +410,15 @@ def _replace_dataloader_init_method() -> Generator[None, None, None]:
     """This context manager is used to add support for re-instantiation of custom (subclasses) of
     :class:`~torch.utils.data.DataLoader`. It patches the ``__init__`` method."""
     classes = _get_all_subclasses(DataLoader) | {DataLoader}
-    wrapped = set()
+    wrapped: set[Any] = set()
     for cls in classes:
-        if cls.__init__ not in wrapped:
+        if cls.__init__ not in wrapped and isinstance(cls, DataLoader):
```
Comment (otaj) [inline src/pytorch_lightning/utilities/data.py:415]: This check is never gonna be true, since we are modifying classes and not instances - therefore `isinstance` call will always return False

File: src/pytorch_lightning/utilities/data.py [hunk 7]
```diff
@@ -408,15 +410,15 @@ def _replace_dataloader_init_method() -> Generator[None, None, None]:
...
             cls._old_init = cls.__init__
             cls.__init__ = _wrap_dataloader_init(cls.__init__)
             wrapped.add(cls.__init__)
     yield
     for cls in classes:
-        if hasattr(cls, "_old_init"):
+        if hasattr(cls, "_old_init") and isinstance(cls, DataLoader):
```
Comment (otaj) [inline src/pytorch_lightning/utilities/data.py:421]: This check is never gonna be true, since we are modifying classes and not instances - therefore `isinstance` call will always return False

File: src/pytorch_lightning/utilities/data.py (hunk 4 above)
Comment (nandwalritik) [reply src/pytorch_lightning/utilities/data.py:139]: `model: Union["pl.LightningModule", "pl.LightningDataModule"]`,  `pl.LightningDataModule` does not have any device attribute. What else can I try?

File: src/pytorch_lightning/utilities/data.py [hunk 8]
```diff
@@ -408,15 +410,15 @@ def _replace_dataloader_init_method() -> Generator[None, None, None]:
     """This context manager is used to add support for re-instantiation of custom (subclasses) of
     :class:`~torch.utils.data.DataLoader`. It patches the ``__init__`` method."""
     classes = _get_all_subclasses(DataLoader) | {DataLoader}
-    wrapped = set()
+    wrapped: set[Any] = set()
```
[the remark this answers is not in the document]
Comment (nandwalritik) [reply src/pytorch_lightning/utilities/data.py:415]: fixed.

File: src/pytorch_lightning/utilities/data.py (hunk 3 above)
Comment (nandwalritik) [reply src/pytorch_lightning/utilities/data.py]: updated to `pl.strategies.Strategy`

**Comment by Borda** [comment]
@nandwalritik @otaj, mind checking the conflicts :otter:

**Comment by otaj** [comment]
Hi, @nandwalritik, do you intend to keep working on this? If not, I can take it over the finish line :zap:

**Comment by Borda** [comment]
> Hi, @nandwalritik, do you intend to keep working on this? If not, I can take it over the finish line zap

let's take it over, pls :rabbit:

**Comment by nandwalritik** [comment]
> Hi, @nandwalritik, do you intend to keep working on this? If not, I can take it over the finish line :zap: 

Sorry got busy with office work, you can take it over @otaj .

**Comment by otaj** [comment]
@Borda It should be finished now, ready for review.
@awaelchli I took the liberty of also typing `lightning_lite/utilities/data.py` since it's very much intertwined.

**Comment by awaelchli** [comment]
@otaj Thank you, and sorry for the merge conflicts when ripping apart some of these utilities. I hope it wasn't too annoying.

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
